src/storage-client/src/controller.rs

```diff
@@ -1728,32 +1727,50 @@ where
 
     Self: StorageController<Timestamp = T>,
 {
-    /// Effectively truncates the `data_shard` associated with `global_id`
-    /// effective as of the system time.
+    /// Reconciles the current state of the collection identified by `id` with
+    /// `updates`.
+    ///
+    /// To do this, for each row, we determine the sum of the diffs in updates
+    /// and the negative of the diffs in the current state. If any updates
+    /// remain, we append them to the identified collection.
+    ///
+    /// Note that using this logic, it's possible to truncate a collection by
+    /// reconciling its current state with an empty set of updates.
     ///
     /// # Panics
     /// - If `id` does not belong to a collection or is not registered as a
     ///   managed collection.
-    async fn truncate_managed_collection(&mut self, id: GlobalId) {
-        let as_of = match self.state.collections[&id]
+    async fn reconcile_managed_collection(&mut self, id: GlobalId, updates: Vec<(Row, Diff)>) {
+        let mut reconciled_updates = HashMap::<Row, Diff>::with_capacity(updates.len());
+
+        for (row, diff) in updates.into_iter() {
+            *reconciled_updates.entry(row).or_default() += diff;
+        }
+
+        match self.state.collections[&id]
             .write_frontier
             .elements()
             .iter()
             .min()
         {
-            Some(f) if f > &T::minimum() => f.step_back().unwrap(),
-            // If collection is closed or the frontier is the minimum, we cannot
-            // or don't need to truncate (respectively).
-            _ => return,
-        };
+            Some(f) if f > &T::minimum() => {
+                let as_of = f.step_back().unwrap();
 
-        let mut negate = self.snapshot(id, as_of).await.unwrap();
+                let negate = self.snapshot(id, as_of).await.unwrap();
 
-        for (_, diff) in negate.iter_mut() {
-            *diff = -*diff;
+                for (row, diff) in negate.into_iter() {
+                    *reconciled_updates.entry(row).or_default() -= diff;
+                }
+            }
+            // If collection is closed or the frontier is the minimum, we cannot
+            // or don't need to truncate (respectively).
+            _ => {}
         }
 
-        self.append_to_managed_collection(id, negate).await;
+        if !reconciled_updates.is_empty() {
+            self.append_to_managed_collection(id, reconciled_updates.into_iter().collect())
```

Comment:
Might we worth tossing a comment that the consolidation isn't really as important as this being a single compare_and_append call to persist. Ditto below